core/src/avm2/globals/flash/display/bitmapdata.rs

```diff
@@ -852,10 +853,249 @@ pub fn rect<'gc>(
 /// Implement `BitmapData.applyFilter`
 pub fn apply_filter<'gc>(
     activation: &mut Activation<'_, 'gc>,
-    _this: Option<Object<'gc>>,
-    _args: &[Value<'gc>],
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
 ) -> Result<Value<'gc>, Error<'gc>> {
-    avm2_stub_method!(activation, "flash.display.BitmapData", "applyFilter");
+    let args = if args.len() == 4 {
+        [
+            args.get(0).expect("Infallible"),
+            args.get(1).expect("Infallible"),
+            args.get(2).expect("Infallible"),
+            args.get(3).expect("Infallible"),
+        ]
+    } else {
+        return Err("ArgumentError: Error #1063: Argument count mismatch on flash.display::BitmapData/applyFilter(). Expected 4, got 0.".into());
+    };
+    if let Some(dest_bitmap) = this.and_then(|this| this.as_bitmap_data()) {
```

Comment:
This should use `as_bitmap_data_wrapper` and `overwrite_cpu_pixels`, like the `draw` method in this file does. This will avoid blocking on any in-progress GPU -> CPU sync.

Reply:
Fixed - but I realized that we can't do this for `source_bitmap`. We don't need the copy to happen here, but we don't want to invalidate it.